experimental/virtio/src/vsock/socket.rs

```diff
@@ -0,0 +1,364 @@
+                |packet| packet.get_dst_port() == dst_port && packet.get_src_port() == src_port,
+                true,
+            )?;
+            self.peer_buffer_size = Wrapping(packet.get_buf_alloc());
+            self.peer_processed_bytes = Wrapping(packet.get_fwd_cnt());
+            // For now we panic if we receive an invalid op.
+            match packet.get_op().expect("Invalid packet received on stream.") {
+                VSockOp::CreditRequest => {
+                    self.send_control_packet(VSockOp::CreditUpdate).unwrap();
+                }
+                VSockOp::CreditUpdate => {
+                    // We already updated our flow-control tracking data, so do nothing.
+                }
+                VSockOp::Request | VSockOp::Response => {
+                    // For now we panic if we receive an invalid op.
+                    panic!("Invalid packet received on stream.");
+                }
+                VSockOp::Rst => {
+                    self.connection_state = ConnectionState::Disconnected;
+                    return None;
+                }
+                VSockOp::Shutdown => {
+                    self.send_control_packet(VSockOp::Rst).unwrap();
+                    self.connection_state = ConnectionState::Disconnected;
+                    return None;
+                }
+                VSockOp::Rw => {
+                    let data = packet.get_payload();
+                    // TODO(#2876): Avoid copying the buffer slice if possible.
+                    let mut result = VecDeque::with_capacity(data.len());
+                    result.extend(data);
+                    return Some(result);
+                }
+            }
+        }
+    }
+
+    /// Tries once to fill the destination with as much data as is currently available, either in
+    /// the pending buffer or from the next available data packet.
+    fn read_partial(&mut self, dest: &mut [u8]) -> Option<usize> {
+        let mut source = match self.pending_data.take() {
+            Some(data) => data,
+            None => self.read_data()?,
+        };
+
+        let len = dest.len();
+        let mut position = 0;
+        while position < len && let Some(byte) = source.pop_front() {
+            dest[position] = byte;
+            position += 1;
+        }
+
+        if !source.is_empty() {
+            self.pending_data.replace(source);
+        }
+        Some(position)
+    }
+}
+
+impl ciborium_io::Read for Socket {
+    type Error = anyhow::Error;
+
+    fn read_exact(&mut self, data: &mut [u8]) -> Result<(), Self::Error> {
+        let len = data.len();
+        let mut count = 0;
+        while count < len {
+            count += self.read_partial(&mut data[count..]).unwrap_or(0);
+        }
+
+        self.processed_bytes += Wrapping(count as u32);
+
+        if self.must_send_credit_update() {
+            self.send_control_packet(VSockOp::CreditUpdate).unwrap();
```

Comment:
Is it possible to return an `Error` here?

Reply:
Changed to return an error for clarity, but it cannot actually happen as the op is not `VSockOp::Rw`.